rust/lance-index/src/vector/pq.rs

```diff
@@ -464,23 +467,63 @@ impl<T: ArrowFloatType + Cosine + Dot + L2 + 'static> ProductQuantizer for Produ
     fn dimension(&self) -> usize {
         self.dimension
     }
+
+    fn codebook_as_fsl(&self) -> FixedSizeListArray {
+        FixedSizeListArray::try_new_from_values(
+            self.codebook.as_ref().clone(),
+            self.dimension as i32,
+        )
+        .unwrap()
+    }
 }
 
 #[allow(clippy::fallible_impl_from)]
-impl From<&dyn ProductQuantizer> for pb::Pq {
-    fn from(pq: &dyn ProductQuantizer) -> Self {
-        Self {
+impl TryFrom<&dyn ProductQuantizer> for pb::Pq {
+    type Error = Error;
+
+    fn try_from(pq: &dyn ProductQuantizer) -> Result<Self> {
+        let fsl = pq.codebook_as_fsl();
+        let tensor = pb::Tensor::try_from(&fsl)?;
```

Comment:
What's the shape of this tensor? Just `[fsl.lenght]`?

Reply:
The shape is `[fsl.length(), fsl.value_length()]` , as tested here

https://github.com/lancedb/lance/pull/1624/files#diff-c908c0060c1d8871846f956e5d6dc0c5ba917d5f5c9059d4bec2b9967116bbb8R527